### PR TITLE
Link redos

### LINKED ISSUE #1425: Regular expression backtracking in linkification
In versions 0.6.0, 0.5.1 and 0.5.0 parsing [this file](https://gist.github.com/blois/bbbed9c5d83d0e7f52ac23d29f21481d#file-matplotlib-pyplt-plot) takes ~2 minutes while it was pretty much instantaneous in 0.3.19.

Steps to reproduce the behavior with smaller repro:

1. [Marked Demo](https://marked.js.org/demo/?text=%20%20dash_capstyle%3A%20%5B%27butt%27%20%7C%20%27round%27%20%7C%20%27projecting%27%5D%0A%20%20dash_joinstyle%3A%20%5B%27miter%27%20%7C%20%27round%27%20%7C%20%27bevel%27%5D%0A%20%20dashes%3A%20sequence%20of%20on%2Foff%20ink%20in%20points%0A%20%20drawstyle%3A%20%5B%27default%27%20%7C%20%27steps%27%20%7C%20%27steps-pre%27%20%7C%20%27steps-mid%27%20%7C%20%27steps-post%27%5D%0A%20%20figure%3A%20a%20%60~.Figure%60%20instance%0A%20%20fillstyle%3A%20%5B%27full%27%20%7C%20%27left%27%20%7C%20%27right%27%20%7C%20%27bottom%27%20%7C%20%27top%27%20%7C%20%27none%27%5D%0A%20%20gid%3A%20an%20id%20string%0A%20%20label%3A%20object%0A%20%20linestyle%20or%20ls%3A%20%5B%27solid%27%20%7C%20%27dashed%27%2C%20%27dashdot%27%2C%20%27dotted%27%20%7C%20(offset%2C%20on-off-dash-seq)%20%7C%20%60%60%27-%27%60%60%20%7C%20%60%60%27--%27%60%60%20%7C%20%60%60%27-.%27%60%60%20%7C%20%60%60%27%3A%27%60%60%20%7C%20%60%60%27None%27%60%60%20%7C%20%60%60%27%20%27%60%60%20%7C%20%60%60%27%27%60%60%5D%0A%20%20linewidth%20or%20lw%3A%20float%20value%20in%20points%0A%20%20marker%3A%20%3Amod%3A%60A%20valid%20marker%20style%20%3Cmatplotlib.markers%3E%60%0A%20%20markeredgecolor%20or%20mec%3A%20any%20matplotlib%20color%0A%20%20markeredgewidth%20or%20mew%3A%20float%20value%20in%20points%0A%20%20markerfacecolor%20or%20mfc%3A%20any%20matplotlib%20color%0A%20%20markerfacecoloralt%20or%20mfcalt%3A%20any%20matplotlib%20color%0A%20%20markersize%20or%20ms%3A%20float%0A%20%20markevery%3A%20%5BNone%20%7C%20int%20%7C%20length-2%20tuple%20of%20int%20%7C%20slice%20%7C%20list%2Farray%20of%20int%20%7C%20float%20%7C%20length-2%20tuple%20of%20float%5D%0A%20%20path_effects%3A%20%60~.AbstractPathEffect%60%0A%20%20picker%3A%20float%20distance%20in%20points%20or%20callable%20pick%20function%20%60%60fn(artist%2C%20event)%60%60%0A%20%20pickradius%3A%20float%20distance%20in%20points%0A%20%20rasterized%3A%20bool%20or%20None%0A%20%20sketch_params%3A%20(scale%3A%20float%2C%20length%3A%20float%2C%20randomness%3A%20float)%0A%20%20snap%3A%20bool%20or%20None%0A%20%20solid_capstyle%3A%20%5B%27butt%27%20%7C%20%27round%27%20%7C%20%20%27projecting%27%5D%0A%20%20solid_joinstyle%3A%20%5B%27miter%27%20%7C%20%27round%27%20%7C%20%27bevel%27%5D%0A%20%20transform%3A%20a%20%3Aclass%3A%60matplotlib.transforms.Transform%60%20instance%0A%20%20url%3A%20a%20url%20string%0A%20%20visible%3A%20bool%0A%20%20xdata%3A%201D%20array%0A%20%20ydata%3A%201D%20array%0A%20%20zorder%3A%20float%0A&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tables%22%3A%20true%2C%0A%20%22xhtml%22%3A%20false%0A%7D)
2. [CommonMark Demo](https://spec.commonmark.org/dingus/?text=%20%20dash_capstyle%3A%20%5B%27butt%27%20%7C%20%27round%27%20%7C%20%27projecting%27%5D%0A%20%20dash_joinstyle%3A%20%5B%27miter%27%20%7C%20%27round%27%20%7C%20%27bevel%27%5D%0A%20%20dashes%3A%20sequence%20of%20on%2Foff%20ink%20in%20points%0A%20%20drawstyle%3A%20%5B%27default%27%20%7C%20%27steps%27%20%7C%20%27steps-pre%27%20%7C%20%27steps-mid%27%20%7C%20%27steps-post%27%5D%0A%20%20figure%3A%20a%20%60~.Figure%60%20instance%0A%20%20fillstyle%3A%20%5B%27full%27%20%7C%20%27left%27%20%7C%20%27right%27%20%7C%20%27bottom%27%20%7C%20%27top%27%20%7C%20%27none%27%5D%0A%20%20gid%3A%20an%20id%20string%0A%20%20label%3A%20object%0A%20%20linestyle%20or%20ls%3A%20%5B%27solid%27%20%7C%20%27dashed%27%2C%20%27dashdot%27%2C%20%27dotted%27%20%7C%20(offset%2C%20on-off-dash-seq)%20%7C%20%60%60%27-%27%60%60%20%7C%20%60%60%27--%27%60%60%20%7C%20%60%60%27-.%27%60%60%20%7C%20%60%60%27%3A%27%60%60%20%7C%20%60%60%27None%27%60%60%20%7C%20%60%60%27%20%27%60%60%20%7C%20%60%60%27%27%60%60%5D%0A%20%20linewidth%20or%20lw%3A%20float%20value%20in%20points%0A%20%20marker%3A%20%3Amod%3A%60A%20valid%20marker%20style%20%3Cmatplotlib.markers%3E%60%0A%20%20markeredgecolor%20or%20mec%3A%20any%20matplotlib%20color%0A%20%20markeredgewidth%20or%20mew%3A%20float%20value%20in%20points%0A%20%20markerfacecolor%20or%20mfc%3A%20any%20matplotlib%20color%0A%20%20markerfacecoloralt%20or%20mfcalt%3A%20any%20matplotlib%20color%0A%20%20markersize%20or%20ms%3A%20float%0A%20%20markevery%3A%20%5BNone%20%7C%20int%20%7C%20length-2%20tuple%20of%20int%20%7C%20slice%20%7C%20list%2Farray%20of%20int%20%7C%20float%20%7C%20length-2%20tuple%20of%20float%5D%0A%20%20path_effects%3A%20%60~.AbstractPathEffect%60%0A%20%20picker%3A%20float%20distance%20in%20points%20or%20callable%20pick%20function%20%60%60fn(artist%2C%20event)%60%60%0A%20%20pickradius%3A%20float%20distance%20in%20points%0A%20%20rasterized%3A%20bool%20or%20None%0A%20%20sketch_params%3A%20(scale%3A%20float%2C%20length%3A%20float%2C%20randomness%3A%20float)%0A%20%20snap%3A%20bool%20or%20None%0A%20%20solid_capstyle%3A%20%5B%27butt%27%20%7C%20%27round%27%20%7C%20%20%27projecting%27%5D%0A%20%20solid_joinstyle%3A%20%5B%27miter%27%20%7C%20%27round%27%20%7C%20%27bevel%27%5D%0A%20%20transform%3A%20a%20%3Aclass%3A%60matplotlib.transforms.Transform%60%20instance%0A%20%20url%3A%20a%20url%20string%0A%20%20visible%3A%20bool%0A%20%20xdata%3A%201D%20array%0A%20%20ydata%3A%201D%20array%0A%20%20zorder%3A%20float%0A)
3. Install marked `npm install --save marked@0.6.0` with the version you are using

**Expected behavior**
Rendering completes in a linear amount of time.

I see that a lot of the time is spent in https://github.com/markedjs/marked/blob/v0.5.1/lib/marked.js#L752

It seems that the regular expression has gotten quite a bit more complex between [v0.3.19](https://regexper.com/#%2F%5E!%3F%5C%5B%28%28%3F%3A%5C%5B%5B%5E%5C%5B%5C%5D%5D*%5C%5D%7C%5C%5C%5B%5C%5B%5C%5D%5D%7C%5B%5E%5C%5B%5C%5D%5D%7C%5C%5D%28%3F%3D%5B%5E%5C%5B%5D*%5C%5D%29%29*%29%5C%5D%5C%28%5Cs*%3C%3F%28%5B%5Cs%5CS%5D*%3F%29%3E%3F%28%3F%3A%5Cs%2B%5B'%22%5D%28%5B%5Cs%5CS%5D*%3F%29%5B'%22%5D%29%3F%5Cs*%5C%29%2F) and [v0.5.1](https://regexper.com/#%2F%5E!%3F%5C%5B%28%28%3F%3A%5C%5B%5B%5E%5C%5B%5C%5D%5D*%5C%5D%7C%5C%5C%5B%5C%5B%5C%5D%5D%3F%7C%60%5B%5E%60%5D*%60%7C%5B%5E%5C%5B%5C%5D%5C%5C%5D%29*%3F%29%5C%5D%5C%28%5Cs*%28%3C%28%3F%3A%5C%5C%5B%3C%3E%5D%3F%7C%5B%5E%5Cs%3C%3E%5C%5C%5D%29*%3E%7C%28%3F%3A%5C%5C%5B%28%29%5D%3F%7C%5C%28%5B%5E%5Cs%5Cx00-%5Cx1f%5C%5C%5D*%5C%29%7C%5B%5E%5Cs%5Cx00-%5Cx1f%28%29%5C%5C%5D%29*%3F%29%28%3F%3A%5Cs%2B%28%22%28%3F%3A%5C%5C%22%3F%7C%5B%5E%22%5C%5C%5D%29*%22%7C%27%28%3F%3A%5C%5C%27%3F%7C%5B%5E%27%5C%5C%5D%29*%27%7C%5C%28%28%3F%3A%5C%5C%5C%29%3F%7C%5B%5E%29%5C%5C%5D%29*%5C%29%29%29%3F%5Cs*%5C%29%2F)

### PR DESCRIPTION
**Pull request by UziTech**
**Marked version:** master

## Description

Fixes redos when code blocks are in square brackets that are not links.

- Fixes #1425 

## Contributor

- [x] Test(s) exist to ensure functionality and minimize regression (if no tests added, list tests covering this PR)

## Committer

In most cases, this should be a different person than the contributor.

- [x] Draft GitHub release notes have been updated.
- [x] CI is green (no forced merge required).
- [x] Merge PR

### CODE REVIEW

**Comment by UziTech** [comment]
/cc @Feder1co5oave @davisjam 

I'm not sure this is a great fix. The test still takes a few hundred milliseconds.

**Comment by Feder1co5oave** [comment]
I tried a different approach but this seems better! It's far from perfect though. I think this issue is inherent in the way [links and code spans bind](https://spec.commonmark.org/0.28/#link-text).

So, for example:

```markdown
[foo`](url)` <-- this is not a link

[`foo]`](url) <-- neither is this

[foo`](url) <-- but this is

[foo`](url)`` code `` <-- and this too.
```

Whenever we try to match a link, if we find a `` ` `` inside the text part, we cannot know if it is a lonely backtick, the start of a codespan contained within the link text, or the start of a codespan which invalidates the link-like part we matched up at that point.
Also, code spans by themselves are hard to parse!:

https://github.com/markedjs/marked/blob/c679e5edb5e5b8adc609928853ef980f06baaf18/lib/marked.js#L546

I think we should consider trying to resemble the codespan matching regex to avoid backtracking too much. Something along the lines of ``(`+) ... \2``...

If you want to merge this fix in the mean time, that's fine too!
